src/Common/src/System/Security/Cryptography/CngPkcs8.cs

```diff
@@ -0,0 +1,336 @@
+                {
+                    AsnReader reader = new AsnReader(manager.Memory, AsnEncodingRules.BER);
+                    int len = reader.GetEncodedValue().Length;
+                    source = source.Slice(0, len);
+
+                    try
+                    {
+                        bytesRead = len;
+                        return ImportPkcs8(source, password);
+                    }
+                    catch (CryptographicException)
+                    {
+                    }
+
+                    ArraySegment<byte> decrypted = KeyFormatHelper.DecryptPkcs8(
+                        password,
+                        manager.Memory.Slice(0, len),
+                        out int innerRead);
+
+                    Span<byte> decryptedSpan = decrypted;
+
+                    try
+                    {
+                        if (innerRead != len)
+                        {
+                            throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding);
+                        }
+
+                        bytesRead = len;
+                        return ImportPkcs8(decryptedSpan);
+                    }
+                    catch (CryptographicException e)
+                    {
+                        throw new CryptographicException(SR.Cryptography_Pkcs8_EncryptedReadFailed, e);
+                    }
+                    finally
+                    {
+                        CryptographicOperations.ZeroMemory(decryptedSpan);
+                        ArrayPool<byte>.Shared.Return(decrypted.Array);
+                    }
+                }
+            }
+        }
+
+        private static AsnWriter RewriteEncryptedPkcs8PrivateKey(
+            AsymmetricAlgorithm key,
+            ReadOnlySpan<byte> passwordBytes,
+            PbeParameters pbeParameters)
+        {
+            Debug.Assert(pbeParameters != null);
+
+            // For RSA:
+            //  * 512-bit key needs ~400 bytes
+            //  * 16384-bit key needs ~10k bytes.
+            //  * KeySize (bits) should avoid re-rent.
+            //
+            // For DSA:
+            //  * 512-bit key needs ~300 bytes.
+            //  * 1024-bit key needs ~400 bytes.
+            //  * 2048-bit key needs ~700 bytes.
+            //  * KeySize (bits) should avoid re-rent.
+            //
+            // For ECC:
+            //  * secp256r1 needs ~200 bytes (named) or ~450 (explicit)
+            //  * secp384r1 needs ~250 bytes (named) or ~600 (explicit)
+            //  * secp521r1 needs ~300 bytes (named) or ~730 (explicit)
+            //  * KeySize (bits) should avoid re-rent for named, and probably
+            //    gets one re-rent for explicit.
+            byte[] rented = ArrayPool<byte>.Shared.Rent(key.KeySize);
+            int rentWritten = 0;
+
+            // If we use 6 bits from each byte, that's 22 * 6 = 132
+            Span<char> randomString = stackalloc char[22];
```

Comment:
It looks like we might be renting below if this is too small?  We could afford a larger stackalloc'd buffer here if it's likely that would avoid renting.

Reply:
This span is just for the random password, as input.   The write-to buffer is always rented (to cover the default size for all keys we'd need ~2kb)